designs/sam_validate.md

```diff
@@ -0,0 +1,58 @@
+# `sam validate` command
+
+This is the design to improve the `sam validate` command to utilise the the `cfn-python-lint` package.
+
+It would be good to go beyond just implementing the CloudFormation package for validation of template files, there is the opportunity to create something better.
+
+## What's the current problem?
+
+The current iteration of `sam validate` only goes so far in validating the the SAM template, it does some cursorary checks:
+
+- Checks `CodeUri` to ensure an S3 URL exists
+- Checks if the template is valid but doesn't return errors
+  - If the template contains other resources outside of SAM it won't validate them
+
+## What will be changed?
+
+We will change how the `sam validate` command works, by integrating the [cfn-lint python package](https://github.com/awslabs/cfn-python-lint/) into the the validation workflow. This change would enhance the user's experience of working with SAM templates.
+
+## Success criteria for the change
+
+The user will be able to find detailed information about their template's validation errors or warnings by running `sam validate`. The validate command will also find any issues in non-SAM resources, such as S3 buckets or SNS topics.
+
+## Out of Scope
+
+- Customisation of the rules through the use of flags
+- Ignoring checks
+- Appending rules
+- Specifying your own validation spec
+
+## User Experience Walkthrough
+
+- Create new project with `sam init --runtime python3.7 foobar`
+- Run `sam validate -t ./template.yaml`
+- Make changes
+- If errors:
+  - Show problems associated with the template
+- If ok:
+  - Awesome, carry on developing
+
+## Implementation
+
+The improved functionality of `sam validate` will utilise the opensource `cfn-python-lint` package (https://github.com/awslabs/cfn-python-lint/) made by AWS. The changes will enhance the existing `sam validate` functionality.
```

Comment:
We should have more details here. Snippets of code or classes could be helpful. Do we need to change any of the template to use [cfn-lint](https://github.com/awslabs/aws-sam-cli/blob/develop/samcli/commands/validate/lib/sam_template_validator.py#L134)? Maybe a small example of how you will configure cfn-lint? Will we need to fetch any data (currently we get [managed policies](https://github.com/awslabs/aws-sam-cli/blob/develop/samcli/commands/validate/lib/sam_template_validator.py#L54))? Should or do we need to configure a logger for the library to print correct debug or warning information?

Reply:
Adding details here.

I still don't quiet understand the changes here. It would be really useful to understand with a snippet or example on how to actually integrate with cfn-lint. The [`sam publish` design](https://github.com/awslabs/aws-sam-cli/blob/develop/designs/sam_publish_cmd.rst#design) is a good example which will help us understand a little more in-depth of what the changes will be. This is especially useful what people reading the design later have little background on the library.